### PR TITLE
Form context

### PR DESCRIPTION
**Pull request by benrudolph**
@czue 
elf: @esoergel

### CODE REVIEW

**Comment by czue** [comment]
(left comments on https://github.com/dimagi/commcare-hq/pull/6667)

File: touchforms/backend/test_xformplayer.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+import touchcare
+
+CUR_DIR = os.path.dirname(__file__)
+
+
+class MockFunctionException(Exception):
+    pass
+
+
+class XFormPlayerTest(unittest.TestCase):
```
Comment (czue) [inline touchforms/backend/test_xformplayer.py]: really awesome that we're able to test this now

File: touchforms/backend/xformplayer.py [hunk 2]
```diff
@@ -115,7 +115,17 @@ def _init(ctx):
 
 
 
-def load_form(xform, instance=None, extensions=None, session_data=None, api_auth=None):
+def load_form(xform, instance=None, extensions=None, session_data=None, api_auth=None, form_context=None):
```
Comment (dimagimon) [inline touchforms/backend/xformplayer.py]: E303 too many blank lines (4)

File: touchforms/backend/xformplayer.py [hunk 3]
```diff
@@ -130,7 +140,7 @@ def load_form(xform, instance=None, extensions=None, session_data=None, api_auth
         preload_data=session_data.get('preloaders', {})
     )
 
-    form.initialize(instance == None, CCInstances(session_data, api_auth))
+    form.initialize(instance == None, CCInstances(session_data, api_auth, form_context))
```
Comment (dimagimon) [inline touchforms/backend/xformplayer.py]: E711 comparison to None should be 'if cond is None:'

File: touchforms/formplayer/static/formplayer/script/xforgasm.js [hunk 4]
```diff
@@ -172,6 +173,7 @@ function WebFormSession(params) {
         var _errMsg = function (msg) {
             return "".concat(ERROR_MESSAGE, msg);
         };
+        params['form_context'] = self.formContext;
```
Comment (dimagimon) [inline touchforms/formplayer/static/formplayer/script/xforgasm.js]: ['form_context'] is better written in dot notation.

File: touchforms/backend/test_xformplayer.py (hunk 1 above)
Comment (benrudolph) [inline touchforms/backend/test_xformplayer.py]: writing these have been really helpful in understanding how TF works too

**Comment by benrudolph** [comment]
this functionality is actually super useful for testing, might be worth it to merge the whole thing even if you don't figure out `all_case_ids` from hq

**Comment by czue** [comment]
@benrudolph I'm cool with that. i'm signed off on the current changes so feel free to self-merge if it's ready to og

**Comment by benrudolph** [comment]
i feel comfortable with it, going to merge.

**Pull request merged**